File: ui/src/routes/api/fleet.containers.ts
import { createFileRoute } from '@tanstack/react-router'
import { json } from '@tanstack/react-start'
import { requireAdmin } from '@/server/api-guard'
import { containerStatus } from '@/server/fleet-docker'
import { db } from '@/server/db/pg'

// GET → container reality per agent (the managed service), admin.
export const Route = createFileRoute('/api/fleet/containers')({
  server: {
    handlers: {
      GET: async ({ request }) => {
        const user = await requireAdmin(request)
        if (user instanceof Response) return user
        const sql = await db()
        const rows = (await sql`select department from agent_defs where enabled order by slug`) as unknown as Array<{
          department: string
        }>
        return json({ containers: await containerStatus(rows.map((r) => r.department)) })
      },
    },
  },
})

File: api/src/routes/fleet/fleet_containers.rs
// GET /api/fleet/containers — port of ui/src/routes/api/fleet.containers.ts.
// Container reality per agent (the managed service), admin.

use crate::error::thrown_internal_error;
use crate::fleet::docker::container_status;
use crate::session::require_admin;
use crate::state::AppState;
use axum::Json;
use axum::extract::State;
use axum::http::HeaderMap;
use axum::response::{IntoResponse, Response};
use serde_json::json;

pub async fn get(State(state): State<AppState>, headers: HeaderMap) -> Response {
    if let Err(gate) = require_admin(&state, &headers).await {
        return gate;
    }
    let rows: Result<Vec<(String,)>, sqlx::Error> =
        sqlx::query_as("select department from agent_defs where enabled order by slug")
            .fetch_all(&state.pg)
            .await;
    let departments = match rows {
        Ok(r) => r.into_iter().map(|(d,)| d).collect::<Vec<_>>(),
        Err(_) => return thrown_internal_error(),
    };
    // containerStatus throws on a docker failure — the whole route 500s the
    // house way (no json body), same as the TS throw.
    match container_status(&departments).await {
        Ok(containers) => Json(json!({ "containers": containers })).into_response(),
        Err(_) => thrown_internal_error(),
    }
}
